pb-rs/src/parser.rs

```diff
@@ -265,6 +265,12 @@ named!(
                     .find(|&&(k, _)| k == "deprecated")
                     .map_or(false, |&(_, v)| str::FromStr::from_str(v)
                         .expect("Cannot parse Deprecated value")),
+                gen_arrayvec: key_vals
+                    .iter()
+                    .find(|&&(k, _)| k == "rust_gen_arrayvec")
```

Comment:
What happens when you run this option through protoc? It looks like you are using this as an option extension without defining it. Is that valid in protos? I haven't tried it yet, but I'll give it a try later this week.

Reply:
I'm not sure what happens with protoc, didn't try it. Since you're asking I assume it's not going to work :) For the custom option business, how do you think that would be idiomatically managed? Presumably we need this code

```protobuf
import "google/protobuf/descriptor.proto";
extend google.protobuf.FieldOptions {
  uint32 rust_gen_arrayvec = 55555; 
}
```
to live somewhere; would it be in a .proto that ships with this library?

(presumably we'd register a number at https://github.com/protocolbuffers/protobuf/blob/master/docs/options.md)

Reply:
LOL. I wasn't sure if it'd work. I thought of some cool things I want to try if it had worked ;)

Error I get:

```
% protoc --python_out=out quick-protobuf/no-std-example/src/no_std.proto
quick-protobuf/no-std-example/src/no_std.proto:17:30: Option "rust_gen_arrayvec" unknown. Ensure that your proto definition file imports the proto which defines the option.
quick-protobuf/no-std-example/src/no_std.proto:19:42: Option "rust_gen_arrayvec" unknown. Ensure that your proto definition file imports the proto which defines the option.
```

I've never registered a number upstream, but it looks like there is a process. I'm also not sure how to include an always-imported dependency. It's possible that we'd need to bundle the  `google/protobuf/descriptor.proto` and another proto with this extension along with pb-rs. It's possible that users of `rust_gen_arrayvec` would need to import our extension file in order to use it. That doesn't feel like the worst API in the world, though if the proto is embedded in pb-rs, the protoc won't work anyway since it won't be able to find the proto with our extension anyway...

Also, to get it to work I needed to change it to this syntax, note the `(...)` around the field option name:

```proto
message NoStdMessage {
  fixed32 num = 1;
  repeated fixed32 nums = 2 [(rust_gen_arrayvec) = 16];
  EmbeddedMessage message = 3;
  repeated EmbeddedMessage messages = 4 [(rust_gen_arrayvec) = 16];
}
```

Reply:
> How do they actually compute the necessary allocation on C side?
I'm not aware of any C implementations that work without a heap, which is why ArrayVec is used here. 

> I think I would prefer some specific pb-rs parameters do deal with it (so we can build a dictionary of message/field arrayvec length, with an optional default length?).

The way I approached this was to add the features necessary to use this in a no_std context, but then allow the user to choose what they want to use on a piecemeal basis. With annotations, you can easily choose to use an ArrayVec for one field and a regular Vec for another, if they so desire. 

The protobuf field option system seems to have been design for exactly this kind of use. I agree it's a little bit distasteful, but I think it's far better than trying to thread field-specific code down from the codegen command line. 

w.r.t. registering the extension: I was thinking it could be a little bit more generic: `(rust.max_length) = 123` could conceivably apply to String (=> ArrayString) or bytes fields.

Reply:
I think the adding the field option is probably considered the "official" way to doing this. protobuf already has a lot of options that are c++ or java specific. That said, we'd need to work out some usability issues to make it work.

If we are going to define a rust specific field / message option extension, then we need some way of importing that options. We could have `pb-rs` special case a `rust-options.proto` import and make sure it is on the import path, but then the protos wouldn't work with protoc. We'd also probably need to get extensions generally working (at least the proto3 version of extensions) in order to parse the base descriptor proto; I haven't tested if pb-rs works with the google descriptor.proto.

IIRC proto3 allows extensions for descriptor options only, whereas proto2 allows extensions to arbitrary messages.

So... while adding the field options is more official, adding command line options to do this would probably be more expedient.

I have a branch that extends the `custom_struct_derive` so you can specify derives for specific messges only using something like `--custom_struct_derive mypackage.MyProtoMessage=Hash,Eq`. I plan on opening a PR with that soon (probably tonight).